File: 顺序刷/动态规划/src/Q8_p416.java
/**
 * 416. 分割等和子集
 * https://leetcode-cn.com/problems/partition-equal-subset-sum/
 *
 * 题目难易：中等
 *
 * 给定一个只包含正整数的非空数组。是否可以将这个数组分割成两个子集，使得两个子集的元素和相等。
 *
 * 注意: 每个数组中的元素不会超过 100 数组的大小不会超过 200
 *
 * 示例 1: 输入: [1, 5, 11, 5] 输出: true 解释: 数组可以分割成 [1, 5, 5] 和 [11].
 *
 * 示例 2: 输入: [1, 2, 3, 5] 输出: false 解释: 数组不能分割成两个元素和相等的子集.
 *
 * 提示：
 *
 * 1 <= nums.length <= 200
 * 1 <= nums[i] <= 100
 */

/**
 * 思路：
 * 爬楼梯角度，数组是步长，数组和的一半是楼顶，问能否到楼顶
 * 0-1背包角度：数组长即物品数，数值即重量，数组和的大小即背包的容量
 * 就是最大重量等于容量
 * 重点
 */
public class Q8_p416 {public static void main(String[] args) {
    int[] nums = {1,5,11,5};
    Solution solution = new Solution();
    boolean ans = solution.canPartition(nums);
    System.out.println(ans);
}

    static class Solution {
        public boolean canPartition(int[] nums) {
            int sum = 0;
            for (int a : nums) {
                sum += a;
            }
            //判断是否能平分
            if (sum % 2 != 0) return false;
            else sum = sum / 2;

            int[][] dp = new int[nums.length][sum+1];
            //初始化dp数组，放入第0个物体后的状态
            for (int j = nums[0]; j <= sum; j++) {
                dp[0][j] = nums[0];
            }
            //递推公式
            for (int i = 1; i < nums.length; i++) {
                for (int j = 1; j <= sum; j++) {
                    //背包容量可以容纳nums[i]
                    if (j >= nums[i]) {
                        dp[i][j] = Math.max(dp[i - 1][j], dp[i - 1][j - nums[i]] + nums[i]);
                    } else {
                        dp[i][j] = dp[i - 1][j];
                    }
                }
            }
            return dp[nums.length-1][sum] == sum;
        }
    }
}

File: python/动态规划/Q8_p416.py
# 416. 分割等和子集
# https://leetcode.cn/problems/partition-equal-subset-sum/description/

# 题目难易：中等

# 给定一个只包含正整数的非空数组。是否可以将这个数组分割成两个子集，使得两个子集的元素和相等。
# 注意: 每个数组中的元素不会超过 100 数组的大小不会超过 200

# 示例 1:

# 输入: [1, 5, 11, 5]
# 输出: true
# 解释: 数组可以分割成 [1, 5, 5] 和 [11].
# 示例 2:

# 输入: [1, 2, 3, 5]
# 输出: false
# 解释: 数组不能分割成两个元素和相等的子集.
# 提示：

# 1 <= nums.length <= 200
# 1 <= nums[i] <= 100
from typing import List

class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        sums = sum(nums)
        if sums % 2 != 0:
            return False
        
        half = sums // 2
        dp = [[0 for _ in range(half + 1)] for _ in range(len(nums))] 
        for i in range(half + 1):
            if i >= nums[0]:
                dp[0][i] = nums[0]
        
        for i in range(len(nums)):
            for j in range(1, half + 1):
                if j - nums[i] >= 0:
                    dp[i][j] = max(dp[i-1][j], dp[i-1][j-nums[i]] + nums[i])
                else:
                    dp[i][j] = dp[i-1][j]
        for i in range(len(nums)):
            if dp[i][half] == half:
                return True
        return False
    
    def canPartition1(self, nums: List[int]) -> bool:
        
        total_sum = sum(nums)

        if total_sum % 2 != 0:
            return False

        target_sum = total_sum // 2
        dp = [[False] * (target_sum + 1) for _ in range(len(nums) + 1)]

        # 初始化第一行（空子集可以得到和为0）
        for i in range(len(nums) + 1):
            dp[i][0] = True

        for i in range(1, len(nums) + 1):
            for j in range(1, target_sum + 1):
                if j < nums[i - 1]:
                    # 当前数字大于目标和时，无法使用该数字
                    dp[i][j] = dp[i - 1][j]
                else:
                    # 当前数字小于等于目标和时，可以选择使用或不使用该数字
                    dp[i][j] = dp[i - 1][j] or dp[i - 1][j - nums[i - 1]]

        return dp[len(nums)][target_sum]

if __name__ == "__main__":
        s = Solution()
        s.canPartition()
